File: Sentinel LDK 7.6/Samples/Admin/vb.net/AdminApiSample.vb
'****************************************************************************
'
' Demo program for Sentinel LDK
'
'
' Copyright (C) 2014, SafeNet, Inc. All rights reserved.
'
'****************************************************************************/

Imports SafeNet.Sentinel
Imports System.Text.RegularExpressions

Module AdminApiSample

    Private adminApi As AdminApi
    Private vendorId As Integer = "37515"

    Sub Main()
        Dim backupdata As String = ""

        Console.WriteLine("A simple demo program for the Sentinel LDK administration functions")
        Console.WriteLine("Copyright (C) 2013, SafeNet, Inc. All rights reserved.")
        Console.WriteLine()

        adminApi = New AdminApi("localhost", 1947, "")

        ''this initializes the adminApi
        sampleAdminConnect()

        ''different samples
        commonSamples(backupdata)
        scopeSamples(vendorId)
        licenseMangerSample()
        keyDescriptionSample()
        restoreData(backupdata)
    End Sub

#Region "samples"

    Sub commonSamples(ByRef backupdata As String)

        ''Retrieve current context
        sampleAdminGet("", "<context></context>")

        ''Retrieve all configuration settings (for backup)
        sampleAdminGet("", _
                    "<admin>" & _
                    " <config>" & _
                    "  <element name=""*""/>" & _
                    " </config>" & _
                    "</admin>", _
                    backupdata _
                    )

        sampleAdminGet("", _
                    "<admin>" & _
                    " <config>" & _
                    "  <element name=""friendlyname"" />" & _
                    " </config>" & _
                    "</admin>" _
                    )

        sampleAdminSet( _
                    "<config>" & _
                    " <enabledetach>1</enabledetach>" & _
                    "</config>" _
                    )

        ''Set configuration defaults and write to ini file
        sampleAdminSet( _
                    "<config>" & _
                    " <set_defaults />" & _
                    " <writeconfig />" & _
                    "</config>" _
                    )

        ''Change some settings and write to ini file"
        ''(For a list of available elements, see reply for \"*\" element below)
        sampleAdminSet( _
                    "<config>" & _
                    " <serveraddrs_clear/>" & _
                    " <serveraddr>10.24.2.16   </serveraddr>" & _
                    " <serveraddr>    10.24.2.16</serveraddr>" & _
                    " <serveraddr>p4p</serveraddr>" & _
                    " <serveraddr>10.24.2.255</serveraddr>" & _
                    " <user_restrictions_clear/>" & _
                    " <user_restriction>deny=baerbel@all</user_restriction>" & _
                    " <access_restrictions_clear />" & _
                    " <access_restriction>deny=10.23.*</access_restriction>" & _
                    " <enabledetach>1</enabledetach>" & _
                    " <writeconfig />" & _
                    "</config>" _
                    )

        ''Retrieve all configuration settings
        sampleAdminGet("", _
                    "<admin>" & _
                    " <config>" & _
                    "  <element name=""*"" />" & _
                    " </config>" & _
                    "</admin>" _
                    )

        ''Add some access restrictions
        sampleAdminSet( _
                    "<config>" & _
                    " <access_restriction>allow=123</access_restriction>" & _
                    " <access_restriction>allow=abcd</access_restriction>" & _
                    " <access_restriction>allow=hello_world</access_restriction>" & _
                    " <writeconfig />" & _
                    "</config>" _
                    )

        ''Retrieve all access restrictions
        sampleAdminGet("", _
                    "<admin>" & _
                    " <config>" & _
                    "  <element name=""access_restriction"" />" & _
                    " </config>" & _
                    "</admin>" _
                    )

        ''Add some more access restrictions
        sampleAdminSet( _
                    "<config>" & _
                    " <access_restriction>allow=more_123</access_restriction>" & _
                    " <access_restriction>allow=more_abcd</access_restriction>" & _
                    " <access_restriction>allow=more_hello_world</access_restriction>" & _
                    " <writeconfig />" & _
                    "</config>" _
                    )

        ''Retrieve all access restrictions
        sampleAdminGet("", _
                    "<admin>" & _
                    " <config>" & _
                    "  <element name=""access_restriction"" />" & _
                    " </config>" & _
                    "</admin>" _
                    )

        ''Delete existing access restrictions and add some new ones
        sampleAdminSet( _
                    "<config>" & _
                    " <access_restrictions_clear/>" & _
                    " <access_restriction>allow=new_123</access_restriction>" & _
                    " <access_restriction>allow=new_abcd</access_restriction>" & _
                    " <writeconfig />" & _
                    "</config>" _
                    )

        ''Retrieve all access restrictions
        sampleAdminGet("", _
                    "<admin>" & _
                    " <config>" & _
                    "  <element name=""access_restriction"" />" & _
                    " </config>" & _
                    "</admin>" _
                    )
    End Sub

    ''using haspscope to retrieve filtered data
    Sub scopeSamples(ByVal vendorId As Integer)

        ''Retrieve some key data for specified vendor (scope with attribute notation)
        ''(for a list of available elements, see reply for \"*\" element below)
        sampleAdminGet( _
                    "<haspscope>" & _
                    " <vendor id=""" & vendorId & """ />" & _
                    "</haspscope>", _
                    "<admin>" & _
                    " <hasp>\n" & _
                    "  <element name=""vendorid"" />" & _
                    "  <element name=""haspid"" />" & _
                    "  <element name=""typename"" />" & _
                    "  <element name=""local"" />" & _
                    "  <element name=""localname"" />" & _
                    " </hasp> " & _
                    "</admin>" _
                    )

        ''Retrieve key data for specified vendor (scope with element notation)
        sampleAdminGet( _
                    "<haspscope>\n" & _
                    " <vendor><id>" & vendorId & "</id></vendor>" & _
                    "</haspscope>", _
                    "<admin>" & _
                    " <hasp>" & _
                    "  <element name=""vendorid"" />" & _
                    "  <element name=""haspid"" />" & _
                    "  <element name=""typename"" />" & _
                    "  <element name=""local"" />" & _
                    "  <element name=""localname"" />" & _
                    " </hasp>" & _
                    "</admin>" _
                    )

        ''Retrieve all product data for specified vendor id"
        sampleAdminGet( _
                    "<haspscope>\n" & _
                    " <vendor><id>" & vendorId & "</id></vendor>" & _
                    "</haspscope>", _
                    "<admin>" & _
                    " <product>" & _
                    "  <element name=""*"" />" & _
                    " </product>" & _
                    "</admin>" _
                    )

        ''Retrieve selected session data for all keys of a specified vendor
        sampleAdminGet( _
                    "<haspscope>\n" & _
                    " <vendor><id>" & vendorId & "</id></vendor>" & _
                    "</haspscope>", _
                    "<admin>" & _
                    " <session>" & _
                    "  <element name=""user"" />" & _
                    "  <element name=""machine"" />" & _
                    "  <element name=""logintime"" />" & _
                    " </session>" & _
                    "</admin>" _
                    )


    End Sub

    ''Sample where the scope uses the key-id
    Sub keyIdSamples(ByVal keyId As Int32)

        ''Retrieve all key data for specified key id"
        sampleAdminGet( _
                    "<haspscope>" & _
                    " <hasp><id>" & keyId & "</id></hasp>" & _
                    "</haspscope>", _
                    "<admin>" & _
                    " <hasp>" & _
                    "  <element name=""*"" />" & _
                    " </hasp>" & _
                    "</admin>" _
                    )


        ''Retrieve all feature data for specified key id"
        sampleAdminGet( _
                    "<haspscope>" & _
                    " <hasp><id>" & keyId & "</id></hasp>" & _
                    "</haspscope>", _
                    "<admin>" & _
                    " <feature>" & _
                    "  <element name=""*"" />" & _
                    " </feature>" & _
                    "</admin>" _
                    )

        ''Retrieve list of current sessions for a specified key
        sampleAdminGet( _
                    "<haspscope>" & _
                    " <hasp><id>" & keyId & "</id></hasp>" & _
                    "</haspscope>", _
                    "<admin>" & _
                    " <session>" & _
                    "  <element name=""*"" />" & _
                    " </session>" & _
                    "</admin>" _
                    )

        ''Retrieve all product data for specified key id"
        sampleAdminGet( _
                    "<haspscope>" & _
                    " <hasp><id>" & keyId & "</id></hasp>" & _
                    "</haspscope>", _
                    "<admin>" & _
                    " <product>" & _
                    "  <element name=""productid"" />" & _
                    "  <element name=""productname"" />" & _
                    "  <element name=""detachable"" />" & _
                    "  <element name=""maxseats"" />" & _
                    "  <element name=""seatsfree"" />" & _
                    " </product>" & _
                    "</admin>" _
                    )


    End Sub

    ''Retrieve License Manager and license related data
    Sub licenseMangerSample()

        ''Retrieve license manager data in XML format (default)
        sampleAdminGet("", _
                   "<admin>" & _
                    " <license_manager>" & _
                    "  <element name=""*"" />" & _
                    " </license_manager>" & _
                    "</admin>", _
                    )


        ''Retrieve license manager data in JSON format
        sampleAdminGet("", _
                    "<admin>" & _
                    " <license_manager format=""json"">" & _
                    "  <element name=""*"" />" & _
                    " </license_manager>" & _
                    "</admin>" _
                    )

        ''Retrieve list of detached licenses
        sampleAdminGet("", _
                    "<admin>" & _
                    " <detached>" & _
                    "  <element name=""*"" />" & _
                    " </detached>" & _
                    "</admin>" _
                    )
    End Sub

    ''Setting a key description (disabled because it was not backed up)
    Sub keyDescriptionSample()
        ''Add a key description (legacy format)
        sampleAdminSet( _
                    "<?xml version=""1.0"" encoding=""UTF-8""?>" & _
                    "<keydescription>" & _
                    " <hasp>" & _
                    "  <id>123456</id>" & _
                    "  <name>One two three four five six</name>" & _
                    " </hasp>" & _
                    "</keydescription>" _
                    )
    End Sub

    ''Examples for uploading of files (disabled because reverting not possible)
    Sub fileUploadSample()

        ''Upload a detach location data file
        sampleAdminSet("file://test_location.xml")

        ''Upload a key names metadata file");
        sampleAdminSet("file://test_key.xml")

        ''Upload a product names metadata file
        sampleAdminSet("file://test_product.xml")

        ''Upload a vendor names metadata file
        sampleAdminSet("file://test_vendor.xml")

        ''Applying V2C
        sampleAdminSet("file://test_update.v2c")

    End Sub

    ''Example how to delete a some sessions
    Sub deleteSessionSample()
        sampleAdminSet( _
                    "<admin>" & _
                    " <deletesession>" & _
                    "  <sessionid>1</sessionid>" & _
                    "  <session id=""2"" />" & _
                    "  <session id=""3"" />" & _
                    "  <sessionid>4</sessionid>" & _
                    " </deletesession>" & _
                    "</admin>" _
                    )

    End Sub

    ''Retrieve list of certificates for specified key (XML)
    Sub certificateSample(ByVal keyId As Int32)
        sampleAdminGet( _
                    "<haspscope>" & _
                    " <hasp><id>" & keyId & "</id></hasp>" & _
                    "</haspscope>", _
                    "<admin>" & _
                    " <certificates>" & _
                    "  <element name=""*"" />" & _
                    " </certificates>" & _
                    "</admin>" _
                    )
    End Sub

#End Region

#Region "wrapper functions"

    Sub sampleAdminGet(ByVal scope As String, ByVal format As String, Optional ByRef target As String = Nothing)

        Dim rc As AdminStatus
        Dim info As String = ""

        Console.WriteLine("sntl_admin_get()")

        rc = adminApi.adminGet(scope, format, info)

        If Not (target Is Nothing) Then
            target = info
        Else
            printState(rc, info)
        End If

    End Sub

    Sub sampleAdminSet(ByVal config As String)

        Dim rc As AdminStatus
        Dim returnStatus As String = ""

        Console.WriteLine("sntl_admin_set()")
        rc = adminApi.adminSet(config, returnStatus)
        printState(rc, returnStatus)

    End Sub

    Sub sampleAdminConnect()
        Dim rc As AdminStatus

        rc = adminApi.connect()
        printState(rc)

    End Sub

#End Region

#Region "help functions"

    Sub restoreData(ByVal backupdata As String)
        Dim config As String = ""
        Dim returnStatus As String = ""

        generateConfigFromOutput(backupdata, config)

        adminApi.adminSet( _
                        "<config>" & _
                        " <serveraddrs_clear/>" & _
                        " <user_restrictions_clear/>" & _
                        " <access_restrictions_clear/>" & _
                        " <writeconfig/>" & _
                        "</config>", returnStatus _
                        )

        adminApi.adminSet(config, returnStatus)

    End Sub

    Sub printState(ByVal status As AdminStatus, Optional ByVal info As String = Nothing)

        If Not (String.IsNullOrEmpty(info)) Then
            Console.WriteLine(info)
        End If

        Console.WriteLine("Result: " & getErrorText(status) & " Statuscode: " & status)
        Console.WriteLine()

    End Sub

    Sub generateConfigFromOutput(ByVal output As String, ByRef config As String)
        If Not (String.IsNullOrEmpty(output)) Then
            Dim regex As New Regex("<admin_status>")
            Dim result() As String = regex.Split(output)

            config = result(0).Replace("</config>", "<writeconfig /></config>")
        End If
    End Sub

    Function getErrorText(ByVal status As AdminStatus) As String
        Select Case status
            Case AdminStatus.StatusOk
                Return "StatusOk"
            Case AdminStatus.InsufMem
                Return "InsufMem"
            Case AdminStatus.InvalidContext
                Return "InvalidContext"
            Case AdminStatus.LmNotFound
                Return "LmNotFound"
            Case AdminStatus.LmTooOld
                Return "LmTooOld"
            Case AdminStatus.BadParameters
                Return "BadParameters"
            Case AdminStatus.LocalNetWorkError
                Return "LocalNetWorkError"
            Case AdminStatus.CannotReadFile
                Return "CannotReadFile"
            Case AdminStatus.ScopeError
                Return "ScopeError"
            Case AdminStatus.PasswordRequired
                Return "PasswordRequired"
            Case AdminStatus.CannotSetPassword
                Return "CannotSetPassword"
            Case AdminStatus.UpdateError
                Return "UpdateError"
            Case AdminStatus.LocalOnly
                Return "LocalOnly"
            Case AdminStatus.BadValue
                Return "BadValue"
            Case AdminStatus.ReadOnly
                Return "ReadOnly"
            Case AdminStatus.ElementUndefined
                Return "ElementUndefined"
            Case AdminStatus.InvalidPointer
                Return "InvalidPointer"
            Case AdminStatus.NoIntegratedLm
                Return "NoIntegratedLm"
            Case AdminStatus.ResultTooBig
                Return "ResultTooBig"
            Case AdminStatus.ScopeResultsEmpty
                Return "ScopeResultsEmpty"
            Case AdminStatus.InvalidVendorCode
                Return "InvalidVendorCode"
            Case AdminStatus.UnknownVendorCode
                Return "UnknownVendorCode"
            Case AdminStatus.DotNetDllBroken
                Return "DotNetDllBroken"
            Case AdminStatus.ConnectMissing
                Return "DotNetDllBroken"
        End Select
        Return ""
    End Function
#End Region

End Module

File: Sentinel LDK 7.6/Samples/Admin/Java/AdminApiSample.java
/****************************************************************************
*
* Demo program for Sentinel LDK
*
*
* Copyright (C) 2014, SafeNet, Inc. All rights reserved.
*
* To use UTF-8 characters in Strings the encoding must be specified 
* at compile time with "-encoding UTF-8".
****************************************************************************/
import SafeNet.AdminApi;
import SafeNet.AdminStatus;

public class AdminApiSample {

    private static AdminApi adminApi;
    private static int vendorId = 37515;

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        String backupdata = "";

        System.out.println("A simple demo program for the Sentinel LDK administration functions");
        System.out.println("Copyright (C) 2013, SafeNet, Inc. All rights reserved.\n");

        adminApi = new AdminApi("localhost", 1947, "");

        //this initializes the adminApi
        sampleAdminConnect();

        //different samples
        backupdata = commonSamples();
        scopeSamples(vendorId);
        licenseMangerSample();
        keyDescriptionSample();
        restoreData(backupdata);
		
		//close the session
		adminApi.dispose();
    }

    private static String commonSamples() {

        //Retrieve current context
        sampleAdminGet("", "<context></context>");

        //Retrieve all configuration settings (for backup)
        String backupdata = sampleAdminGet("",
                "<admin>"
                + " <config>"
                + "  <element name=\"*\"/>"
                + " </config>"
                + "</admin>");

        sampleAdminGet("",
                "<admin>"
                + " <config>"
                + "  <element name=\"friendlyname\" />"
                + " </config>"
                + "</admin>");

        sampleAdminSet(
                "<config>"
                + " <enabledetach>1</enabledetach>"
                + "</config>");

        //Set configuration defaults and write to ini file
        sampleAdminSet(
                "<config>"
                + " <set_defaults />"
                + " <writeconfig />"
                + "</config>");

        //Change some settings and write to ini file"
        //(For a list of available elements, see reply for \"*\" element below)
        sampleAdminSet(
                "<config>"
                + " <serveraddrs_clear/>"
                + " <serveraddr>10.24.2.16   </serveraddr>"
                + " <serveraddr>    10.24.2.16</serveraddr>"
                + " <serveraddr>p4p</serveraddr>"
                + " <serveraddr>10.24.2.255</serveraddr>"
                + " <user_restrictions_clear/>"
                + " <user_restriction>deny=baerbel@all</user_restriction>"
                + " <access_restrictions_clear />"
                + " <access_restriction>deny=10.23.*</access_restriction>"
                + " <enabledetach>1</enabledetach>"
                + " <writeconfig />"
                + "</config>");

        //Retrieve all configuration settings
        sampleAdminGet("",
                "<admin>"
                + " <config>"
                + "  <element name=\"*\" />"
                + " </config>"
                + "</admin>");

        //Add some access restrictions
        sampleAdminSet(
                "<config>"
                + " <access_restriction>allow=123</access_restriction>"
                + " <access_restriction>allow=abcd</access_restriction>"
                + " <access_restriction>allow=hello_world</access_restriction>"
                + " <writeconfig />"
                + "</config>");

        //Retrieve all access restrictions
        sampleAdminGet("",
                "<admin>"
                + " <config>"
                + "  <element name=\"access_restriction\" />"
                + " </config>"
                + "</admin>");

        //Add some more access restrictions
        sampleAdminSet(
                "<config>"
                + " <access_restriction>allow=more_123</access_restriction>"
                + " <access_restriction>allow=more_abcd</access_restriction>"
                + " <access_restriction>allow=more_hello_world</access_restriction>"
                + " <writeconfig />"
                + "</config>");

        //Retrieve all access restrictions
        sampleAdminGet("",
                "<admin>"
                + " <config>"
                + "  <element name=\"access_restriction\" />"
                + " </config>"
                + "</admin>");

        //Delete existing access restrictions and add some new ones
        sampleAdminSet(
                "<config>"
                + " <access_restrictions_clear/>"
                + " <access_restriction>allow=new_123</access_restriction>"
                + " <access_restriction>allow=new_abcd</access_restriction>"
                + " <writeconfig />"
                + "</config>");

        //Retrieve all access restrictions
        sampleAdminGet("",
                "<admin>"
                + " <config>"
                + "  <element name=\"access_restriction\" />"
                + " </config>"
                + "</admin>");

        return backupdata;
    }

    //using haspscope to retrieve filtered data
    private static void scopeSamples(int vendorId) {

        //Retrieve some key data for specified vendor (scope with attribute notation)
        //(for a list of available elements, see reply for \"*\" element below)
        sampleAdminGet(
                "<haspscope>"
                + " <vendor id=\"" + vendorId + "\" />"
                + "</haspscope>", "<admin>"
                + " <hasp>\\n"
                + "  <element name=\"vendorid\" />"
                + "  <element name=\"haspid\" />"
                + "  <element name=\"typename\" />"
                + "  <element name=\"local\" />"
                + "  <element name=\"localname\" />"
                + " </hasp> "
                + "</admin>");

        //Retrieve key data for specified vendor (scope with element notation)
        sampleAdminGet(
                "<haspscope>\\n"
                + " <vendor><id>" + vendorId + "</id></vendor>"
                + "</haspscope>", "<admin>"
                + " <hasp>"
                + "  <element name=\"vendorid\" />"
                + "  <element name=\"haspid\" />"
                + "  <element name=\"typename\" />"
                + "  <element name=\"local\" />"
                + "  <element name=\"localname\" />"
                + " </hasp>"
                + "</admin>");

        //Retrieve all product data for specified vendor id"
        sampleAdminGet(
                "<haspscope>\\n"
                + " <vendor><id>" + vendorId + "</id></vendor>"
                + "</haspscope>", "<admin>"
                + " <product>"
                + "  <element name=\"*\" />"
                + " </product>"
                + "</admin>");

        //Retrieve selected session data for all keys of a specified vendor
        sampleAdminGet(
                "<haspscope>\\n"
                + " <vendor><id>" + vendorId + "</id></vendor>"
                + "</haspscope>", "<admin>"
                + " <session>"
                + "  <element name=\"user\" />"
                + "  <element name=\"machine\" />"
                + "  <element name=\"logintime\" />"
                + " </session>"
                + "</admin>");


    }

    //Sample where the scope uses the key-id
    private static void keyIdSamples(int keyId) {
        
        //Retrieve all key data for specified key id"
        sampleAdminGet(
                "<haspscope>"
                + " <hasp><id>" + keyId + "</id></hasp>"
                + "</haspscope>", "<admin>"
                + " <hasp>"
                + "  <element name=\"*\" />"
                + " </hasp>"
                + "</admin>");

        //Retrieve all feature data for specified key id"
        sampleAdminGet(
                "<haspscope>"
                + " <hasp><id>" + keyId + "</id></hasp>"
                + "</haspscope>", "<admin>"
                + " <feature>"
                + "  <element name=\"*\" />"
                + " </feature>"
                + "</admin>");

        //Retrieve list of current sessions for a specified key
        sampleAdminGet(
                "<haspscope>"
                + " <hasp><id>" + keyId + "</id></hasp>"
                + "</haspscope>", "<admin>"
                + " <session>"
                + "  <element name=\"*\" />"
                + " </session>"
                + "</admin>");

        //Retrieve all product data for specified key id"
        sampleAdminGet(
                "<haspscope>"
                + " <hasp><id>" + keyId + "</id></hasp>"
                + "</haspscope>", "<admin>"
                + " <product>"
                + "  <element name=\"productid\" />"
                + "  <element name=\"productname\" />"
                + "  <element name=\"detachable\" />"
                + "  <element name=\"maxseats\" />"
                + "  <element name=\"seatsfree\" />"
                + " </product>"
                + "</admin>");
    }

    //Retrieve License Manager and license related data
    private static void licenseMangerSample() {
        
        //Retrieve license manager data in XML format (default)
        sampleAdminGet("",
                "<admin>"
                + " <license_manager>"
                + "  <element name=\"*\" />"
                + " </license_manager>"
                + "</admin>");

        //Retrieve license manager data in JSON format
        sampleAdminGet("",
                "<admin>"
                + " <license_manager format=\"json\">"
                + "  <element name=\"*\" />"
                + " </license_manager>"
                + "</admin>");

        //Retrieve list of detached licenses
        sampleAdminGet("",
                "<admin>"
                + " <detached>"
                + "  <element name=\"*\" />"
                + " </detached>"
                + "</admin>");
    }

    //Setting a key description (disabled because it was not backed up)
    private static void keyDescriptionSample() {
        
        //Add a key description (legacy format)
        sampleAdminSet(
                "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
                + "<keydescription>"
                + " <hasp>"
                + "  <id>123456</id>"
                + "  <name>One two three four five six</name>"
                + " </hasp>"
                + "</keydescription>");
    }

    //Examples for uploading of files (disabled because reverting not possible)
    private static void fileUploadSample() {
        
        //Upload a detach location data file
        sampleAdminSet("file://test_location.xml");

        //Upload a key names metadata file");
        sampleAdminSet("file://test_key.xml");

        //Upload a product names metadata file
        sampleAdminSet("file://test_product.xml");

        //Upload a vendor names metadata file
        sampleAdminSet("file://test_vendor.xml");

        //Applying V2C
        sampleAdminSet("file://test_update.v2c");
    }

    //Example how to delete a some sessions
    private static void deleteSessionSample() {
        sampleAdminSet(
                "<admin>"
                + " <deletesession>"
                + "  <sessionid>1</sessionid>"
                + "  <session id=\"2\" />"
                + "  <session id=\"3\" />"
                + "  <sessionid>4</sessionid>"
                + " </deletesession>"
                + "</admin>");
    }

    //Retrieve list of certificates for specified key (XML)
    private static void certificateSample(int keyId) {
        sampleAdminGet(
                "<haspscope>"
                + " <hasp><id>" + keyId + "</id></hasp>"
                + "</haspscope>", "<admin>"
                + " <certificates>"
                + "  <element name=\"*\" />"
                + " </certificates>"
                + "</admin>");
    }

    private static String sampleAdminGet(String scope, String format) {
        int rc = 0;
        String info = "";

        System.out.println("sntl_admin_get()");

        info = adminApi.getConfig(scope, format);
        rc = adminApi.getLastApiError();
        printState(rc, info);

        return info;
    }

    private static void sampleAdminSet(String config) {
        int rc = 0;
        String returnStatus = "";

        System.out.println("sntl_admin_set()");
        
        returnStatus = adminApi.setConfig(config);
        rc = adminApi.getLastApiError();
        printState(rc, returnStatus);
    }

    private static void sampleAdminConnect() {
        int rc = 0;
        
        rc = adminApi.connect();
        printState(rc);
    }

    private static void restoreData(String backupdata) {
        String config = generateConfigFromOutput(backupdata);

        adminApi.setConfig(
                "<config>"
                + " <serveraddrs_clear/>"
                + " <user_restrictions_clear/>"
                + " <access_restrictions_clear/>"
                + " <writeconfig/>"
                + "</config>");

        adminApi.setConfig(config);
    }

    private static void printState(int status) {
        printState(status, null);
    }

    private static void printState(int status, String info) {
        if (!(info == null || ("").equals(info))) {
            System.out.println(info);
        }

        System.out.println("Result: " + getErrorText(status) + " Statuscode: " + status + "\n");
    }

    private static String generateConfigFromOutput(String output) {
        String[] result = output.split("<admin_status");
        return result[0].replace("</config>", "<writeconfig /></config>");
    }

    private static String getErrorText(int status) {
        switch (status) {
            case AdminStatus.SNTL_ADMIN_STATUS_OK:
                return "StatusOk";
            case AdminStatus.SNTL_ADMIN_LM_NOT_FOUND:
                return "LmNotFound";
            case AdminStatus.SNTL_ADMIN_LM_TOO_OLD:
                return "LmTooOld";
            case AdminStatus.SNTL_ADMIN_DLL_BROKEN:
                return "DllBroken";
            case AdminStatus.SNTL_ADMIN_CONNECT_MISSING:
                return "ConnectMissing";
            case AdminStatus.SNTL_ADMIN_BAD_PARAMETERS:
                return "BadParameters";
            case AdminStatus.SNTL_ADMIN_BAD_VALUE:
                return "BadValue";
        }
        return "";
    }
}
